marshal.go

```diff
@@ -133,49 +134,74 @@ func marshalImpl(vo reflect.Value, w io.Writer, options *MarshalOptions) error {
 
 			for _, vn := range lst {
 				// Write element ID
-				if _, err := w.Write(e.b); err != nil {
-					return err
+				var headerSize uint64
+				if n, err := w.Write(e.b); err != nil {
+					return pos, err
+				} else {
+					headerSize += uint64(n)
 				}
 				var bw io.Writer
 				if unknown {
 					// Directly write length unspecified element
 					bsz := encodeDataSize(uint64(sizeUnknown), 0)
-					if _, err := w.Write(bsz); err != nil {
-						return err
+					if n, err := w.Write(bsz); err != nil {
+						return pos, err
+					} else {
+						headerSize += uint64(n)
 					}
 					bw = w
 				} else {
 					bw = &bytes.Buffer{}
 				}
 
+				elem := &Element{
+					Value:    vn.Interface(),
+					Name:     tag.name,
+					Position: pos,
+					Size:     sizeUnknown,
+					Parent:   parent,
+				}
+
+				var size uint64
 				if e.t == TypeMaster {
-					if err := marshalImpl(vn, bw, options); err != nil {
-						return err
+					if p, err := marshalImpl(vn, bw, pos+headerSize, elem, options); err != nil {
+						return pos, err
+					} else {
+						size = p - pos - headerSize
 					}
 				} else {
 					bc, err := perTypeEncoder[e.t](vn.Interface(), tag.size)
 					if err != nil {
-						return err
+						return pos, err
 					}
-					if _, err := bw.Write(bc); err != nil {
-						return err
+					if n, err := bw.Write(bc); err != nil {
+						return pos, err
+					} else {
+						size = uint64(n)
 					}
 				}
 
 				// Write element with length
 				if !unknown {
-					bsz := encodeDataSize(uint64(bw.(*bytes.Buffer).Len()), options.dataSizeLen)
-					if _, err := w.Write(bsz); err != nil {
-						return err
+					elem.Size = size
+					bsz := encodeDataSize(elem.Size, options.dataSizeLen)
+					if n, err := w.Write(bsz); err != nil {
+						return pos, err
+					} else {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)